xrootd/encoder/encoder.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package encoder // import "go-hep.org/x/hep/xrootd/encoder"
+
+import (
+	"encoding/binary"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// MarshalRequest encodes request body alongside with request and stream ids
+func MarshalRequest(requestID uint16, streamID protocol.StreamID, requestBody interface{}) ([]byte, error) {
+	requestHeader := make([]byte, 4)
+
+	copy(requestHeader, streamID[:])
+	binary.BigEndian.PutUint16(requestHeader[2:], requestID)
+
+	b, err := Marshal(requestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(requestHeader, b...), nil
+}
+
+// Marshal encodes structure to the bytes following the XRootd protocol specification.
+// Fields are encoded in the same order as they are specified in the struct definition.
+// Each field is encoded in network byte order (BigEndian) without any alignment or padding.
+// Slices and arrays of uint8 are binary copied without further encoding.
+// Supported types are: uint8, uint16, int32, int64, slices and arrays of uint8.
+func Marshal(x interface{}) ([]byte, error) {
+	v := reflect.ValueOf(x)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+
+	if v.Kind() != reflect.Struct {
+		return nil, errors.Errorf("Cannot marshal %s, struct is expected", v.Kind())
+	}
+
+	dataSize, err := calculateSizeForMarshaling(v)
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, dataSize)
+	pos := 0
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
```

Comment:
instead of switching on the field `Kind`, I would directly `type-switch` on its interface value:

```go
switch fv := field.Interface().(type) {
case uint8:
    fieldSize = 1
    data[pos] = fv
case uint16:
    fieldSize = 2
    binary.BigEndian.PutUint16(data[pos:pos+fieldSize], fv))
...
case []byte:
    fieldSize = copy(data[pos:], fv)
default:
    if field.Kind() == reflect.Array {
        // handle it
    } else {
        errors.Errorf("encoder: can not Marshal invalid field type %T", fv)
    }
}
```

le

Reply:
If I have done it correctly, it is 1.5x - 2x slower: 
1130 ns/op vs 1888 ns/op for Marshal and 
733 ns/op vs 1780 ns/op for Unmarshal. 

https://github.com/EgorMatirov/hep/commit/104d3630761afefc267392a1c9fefa70ef2fd1b4

Reply:
ok. that's interesting...
let's use `reflect.Value.Kind` then (it should be a moot point once we get per-type (un)marshaling)